src/streamlink_cli/utils/formatter.py

```diff
@@ -1,63 +1,12 @@
-from string import Formatter as StringFormatter
-from typing import Any, Callable, Dict, Optional
+from typing import Dict, Optional
 
+from streamlink.utils import Formatter as _BaseFormatter
 from streamlink_cli.utils.path import replace_chars
 
 
-# we only need string.Formatter for calling its parse() method, which returns `_string.formatter_parser(string)`.
-_stringformatter = StringFormatter()
-
-
-def _identity(obj):
-    return obj
-
-
-class Formatter:
-    def __init__(self, mapping: Dict[str, Callable], formatting: Optional[Dict[str, Callable]] = None):
-        super().__init__()
-        self.mapping = mapping
-        self.formatting = formatting or {}
-        self.cache = {}
-
-    def _get_value(self, field_name: str, format_spec: str, defaults: Dict[str, str]) -> Any:
-        if field_name not in self.mapping:
-            return defaults.get(field_name, f"{{{field_name}}}" if not format_spec else f"{{{field_name}:{format_spec}}}")
-
-        if field_name in self.cache:
-            value = self.cache[field_name]
-        else:
-            value = self.mapping[field_name]()
-            self.cache[field_name] = value
-
-        if value is None:
-            value = defaults.get(field_name, "")
-
-        if format_spec and field_name in self.formatting:
-            # noinspection PyBroadException
-            try:
-                return self.formatting[field_name](value, format_spec)
-            except Exception:
-                return f"{{{field_name}:{format_spec}}}"
-
-        return value
-
-    def _format(self, string: str, mapper: Callable[[str], str], defaults: Dict[str, str]) -> str:
-        result = []
-
-        for literal_text, field_name, format_spec, conversion in _stringformatter.parse(string):
-            if literal_text:
-                result.append(literal_text)
-
-            if field_name is None:
-                continue
-
-            value = self._get_value(field_name, format_spec, defaults)
-            result.append(mapper(str(value)))
-
-        return "".join(result)
-
+class Formatter(_BaseFormatter):
```

Comment:
two `Formatter` with the same name seems suboptimal,
might just remove this file and use only `src/streamlink/utils/formatter.py`

Reply:
The reason for this are the `title` and `filename` methods on the subclass which keep the original interface for the CLI stuff. `filename` depends on `streamlink_cli.utils.path` which has the logic for file name substitutions and belongs into the CLI module.